### ISSUE TITLE
Unexpected error when using annotated types

### ISSUE THREAD

**Issue by erszcz** [josefs/Gradualizer#224]
I'm experiencing a strange type error with the following code:

```erlang
    66	-spec handle_purge_message_form(From :: jid:jid(), ArcJID :: jid:jid(),
    67	                                IQ :: jlib:iq()) ->
    68	    jlib:iq() | ignore | {error, term(), jlib:iq()}.
    69	handle_purge_message_form(#jid{} = From, #jid{} = ArcJID,
    70	                          #iq{xmlns=_MamNs, sub_el = QueryEl} = IQ) ->
    71	    ...
```

The error is:

```
The record patterns for record #jid on line 69 at column 27 is expected to have type From :: jid:jid().
```

When I modify the spec not to use the type variable `From`:

```erlang
-spec handle_purge_message_form(From :: jid:jid(), ArcJID :: jid:jid(),
                                IQ :: jlib:iq()) ->
    ...
```

the error is gone, but immediately an identical one for `ArcJID` pops up for the same line. Same happens for `IQ` when I proceed. Finally, no errors for line 69 are reported once the spec is free from type vars:

```erlang
-spec handle_purge_message_form(jid:jid(), jid:jid(), jlib:iq()) ->
    ...
```

This seems completely unexpected to me. Might there be a bug in some type equivalence checks between types named with a variable and same types, but with no explicit names attached?

**Comment by zuiderkwast** [comment]
Seems to fail on the handling of annotated types, yes. Thanks for reporting the bug!

Feel free to add a minimal `known_problems` test case!

(For the solution, I think these should be removed in the normalization, which perhaps isn't happening where it should in some case.)

**Comment by erszcz** [comment]
I've minimised the problem to:

```erlang
-record(r, {}).
-type r() :: #r{}.

-spec named_record_arg(R :: r()) -> {ok, R}.
named_record_arg(#r{} = R) -> {ok, R}.
```

Interestingly, the problem disappears when we rewrite it as:

```erlang
-spec named_record_arg(R) -> {ok, R} when
      R :: r().
named_record_arg(#r{} = R) -> {ok, R}.
```

[Here's the test case](https://github.com/josefs/Gradualizer/pull/225).

**Comment by NelsonVides** [comment]
> Interestingly, the problem disappears when we rewrite it as:
> 
> ```erlang
> -spec named_record_arg(R) -> {ok, R} when
>       R :: r().
> named_record_arg(#r{} = R) -> {ok, R}.
> ```
The reason why that problem disappears is because the erlang compiler gives completely different ASTs to those two functions. I actually always thought that the `when` keyword would be nothing else than syntactic sugar for annotated types, but it turns out it is a different thing. Here's the output of `epp:parse_file/2` over these functions:

```erlang
-record(r, {}).
-type r() :: #r{}.

-spec annotated_named_record_arg(R :: r()) -> {ok, R :: r()}.
annotated_named_record_arg(#r{} = R) -> {ok, R}.

-spec bounded_named_record_arg(R) -> {ok, R} when R :: r().
bounded_named_record_arg(#r{} = R) -> {ok, R}.
```

produces

```erlang
     {attribute,9,spec,
                {{annotated_named_record_arg,1},
                 [{type,9,'fun',
                        [{type,9,product,
                               [{ann_type,9,[{var,9,'R'},{user_type,9,r,[]}]}]},
                         {type,9,tuple,
                               [{atom,9,ok},
                                {ann_type,9,[{var,9,'R'},{user_type,9,r,[]}]}]}]}]}},
     {function,10,annotated_named_record_arg,1,
               [{clause,10,
                        [{match,10,{record,10,r,[]},{var,10,'R'}}],
                        [],
                        [{tuple,10,[{atom,10,ok},{var,10,'R'}]}]}]},

     {attribute,12,spec,
                {{bounded_named_record_arg,1},
                 [{type,12,bounded_fun,
                        [{type,12,'fun',
                               [{type,12,product,[{var,12,'R'}]},
                                {type,12,tuple,[{atom,12,ok},{var,12,'R'}]}]},
                         [{type,12,constraint,
                                [{atom,12,is_subtype},
                                 [{var,12,'R'},{user_type,12,r,[]}]]}]]}]}},
     {function,13,bounded_named_record_arg,1,
               [{clause,13,
                        [{match,13,{record,13,r,[]},{var,13,'R'}}],
                        [],
                        [{tuple,13,[{atom,13,ok},{var,13,'R'}]}]}]},
     {eof,14}]}
```

So we can see that, while the AST of the function is exactly the same, the AST of the specs are quite different, one using annotated types and the other using bounded funs with constraints. Mind-blowing to me.

And as Gradualizer checks the AST, these two checks will follow totally different paths 🤷‍♂

**Comment by zuiderkwast** [comment]
I have generated examples of all of Erlang's AST here: https://github.com/zuiderkwast/erlang_abstract_format, for easy reference.

**Comment by zuiderkwast** [comment]
Note that these are not type variables, because this doesn't even compile:

```Erlang
-spec f(A) -> A :: list().
f(A) -> lists:sort(A) ++ A.
```

The compiler gives the error `type variable 'A' is only used once (is unbound)` for this. The first `A` is the type variable which only is used once. The second `A` is not a type variable but just a "name" of the return value which has type `list()`.

**Comment by zuiderkwast** [comment]
Fixed by #228
